### PR TITLE
Feature/validation and rules issues

### LINKED ISSUE #24: Required validator not working for NaN
When using the "useValidation" hook for validating model with a property that is NaN using required as validator, the validation passes.

### LINKED ISSUE #30: Explicit test error message
Hello,

When a validator or a rule is undefined  in a test suite, the error in console is "**Cannot read property 'runReader' of undefined."**

An explicit error message will be useful.

### LINKED ISSUE #31: Missing always valid (empty) validator
Usefull when starting a component and you do not have yet any validation rules

### LINKED ISSUE #32: HoV shape doesn't check if property exist
Hello,

I have the following model to validate:
`const model = {
flowDictionary: V.shape({
        name: V.required
    })
}`

In some cases, my model is an empty object. In this situation, the validation for **flowDictionary**   passes even though property that is required isn't set.
The workaround in this situation is that i set my model from the beginning with property **flowDictionary: {}**

Maybe HoV shape should check first if the property exist. If not, the validation should fail.

Thanks!

### PR DESCRIPTION
**Pull request by fraliv13**

### CODE REVIEW

File: packages/change-tracking-react/src/lensProxy/index.d.ts [hunk 1]
```diff
@@ -9,13 +9,13 @@ export function eject<TValue>(lens: LensProxy<TValue>): ProfunctorState<TValue>;
  * Sets a new value in the profunctor state
  * @see https://github.com/osstotalsoft/jsbb/tree/master/packages/change-tracking-react/src/lensProxy/README.md#set
  */
-export function set<TValue, TResult>(lens: LensProxy<TValue>, newValue: TResult): void;
+export function set(lens: LensProxy<any>, newValue: any): void;
```
Comment (oncicaradupopovici) [inline packages/change-tracking-react/src/lensProxy/index.d.ts:12]: Why any?

File: packages/pure-validations/src/__tests__/composedValidators.tests.js [hunk 2]
```diff
@@ -375,4 +375,122 @@ describe("composed validators:", () => {
     // Assert
     expect(validation).toStrictEqual(Failure(ValidationError([], { 0: ValidationError(["Wrong", "Too short"]) })));
   });
+
+  it("required on missing field: ", () => {
+    // Arrange
+    const model = {
+    };
+
+    const validator = stopOnFirstFailure(shape({ name: required }));
```
Comment (oncicaradupopovici) [inline packages/pure-validations/src/__tests__/composedValidators.tests.js]: Why stopOnFirstFailure?

File: packages/pure-validations/src/__tests__/composedValidators.tests.js [hunk 3]
```diff
@@ -375,4 +375,122 @@ describe("composed validators:", () => {
...
+        })
+      )
+    );
+  });
+
+  it("required on a field in a missing nested object: ", () => {
+    // Arrange
+    const model = undefined;
+
+    const validator = stopOnFirstFailure(shape({parent: shape({ name: required })}));
```
Comment (oncicaradupopovici) [inline packages/pure-validations/src/__tests__/composedValidators.tests.js]: Again why stopOnFirstFailure?

File: packages/pure-validations/src/__tests__/composedValidators.tests.js [hunk 4]
```diff
@@ -375,4 +375,122 @@ describe("composed validators:", () => {
...
+        })
+      )
+    );
+  });
+
+  it("required on a field in a missing object: ", () => {
+    // Arrange
+    const model = undefined;
+
+    const validator = stopOnFirstFailure(shape({ name: required }));
```
Comment (oncicaradupopovici) [inline packages/pure-validations/src/__tests__/composedValidators.tests.js]: Again why stopOnFirstFailure?

File: packages/change-tracking-react/src/lensProxy/index.d.ts (hunk 1 above)
Comment (fraliv13) [reply packages/change-tracking-react/src/lensProxy/index.d.ts:12]: dts lint reports errors for this situation: 
https://github.com/Microsoft/dtslint/blob/master/docs/no-unnecessary-generics.md

File: packages/pure-validations/src/__tests__/composedValidators.tests.js (hunk 2 above)
Comment (fraliv13) [reply packages/pure-validations/src/__tests__/composedValidators.tests.js]: I removed it

File: packages/pure-validations/src/__tests__/composedValidators.tests.js (hunk 4 above)
Comment (fraliv13) [reply packages/pure-validations/src/__tests__/composedValidators.tests.js]: I removed it

File: packages/pure-validations/src/__tests__/composedValidators.tests.js (hunk 3 above)
Comment (fraliv13) [reply packages/pure-validations/src/__tests__/composedValidators.tests.js]: I removed it

**Pull request merged**